### ISSUE TITLE
Rancher Load Balancer Healthchecks causing a lot of (104: Connection reset by peer)

### ISSUE THREAD

**Issue by suizman** [rancher/rancher#5464]
**Rancher Version:**
v1.1.0

**Docker Version:**
1.11.2

**OS and where are the hosts located? (cloud, bare metal, etc):**
AWS Red Hat Enterprise Linux Server release 7.2 (Maipo)

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Non HA Rancher server + AWS Aurora db + 40~ AWS autoscale group nodes

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**

Run some app with docker nginx:1.10.0 create rancher LB on top of it.

Nginx config: 

```
user root root;
worker_processes  1;

pid        /var/run/nginx.pid;


events {
    worker_connections  1024;
}


http {
    include       mime.types;

    sendfile        on;
    keepalive_timeout  65;

    server {
        listen       80;
        server_name  localhost;

        location / {
            root   /var/www;
            index  index.html index.htm;
        }
        access_log /var/log/nginx/access.log;
        error_log /var/log/nginx/error.log  info;

        error_page  404     /index.html;

        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   /var/www;
        }

    }

}
```

**Results:** 

```
18/7/2016 17:14:4210.42.89.204 - - [18/Jul/2016:15:14:42 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:422016/07/18 15:14:42 [info] 8#8: *3240 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.89.204, server: localhost, request: "GET /api/heartbeat HTTP/1.0"
18/7/2016 17:14:4210.42.12.39 - - [18/Jul/2016:15:14:42 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:4310.42.226.193 - - [18/Jul/2016:15:14:43 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:432016/07/18 15:14:43 [info] 8#8: *3242 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.226.193, server: localhost, request: "GET /api/heartbeat HTTP/1.0"
18/7/2016 17:14:442016/07/18 15:14:44 [info] 8#8: *3243 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.224.205, server: localhost, request: "GET /api/heartbeat HTTP/1.0"
18/7/2016 17:14:4410.42.224.205 - - [18/Jul/2016:15:14:44 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:442016/07/18 15:14:44 [info] 8#8: *3244 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.89.204, server: localhost, request: "GET /api/heartbeat HTTP/1.0"
18/7/2016 17:14:4410.42.89.204 - - [18/Jul/2016:15:14:44 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:4410.42.12.39 - - [18/Jul/2016:15:14:44 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:452016/07/18 15:14:45 [info] 8#8: *3246 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.226.193, server: localhost, request: "GET /api/heartbeat HTTP/1.0"
18/7/2016 17:14:4510.42.226.193 - - [18/Jul/2016:15:14:45 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:4610.42.224.205 - - [18/Jul/2016:15:14:46 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:4610.42.89.204 - - [18/Jul/2016:15:14:46 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:4610.42.12.39 - - [18/Jul/2016:15:14:46 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:4710.42.226.193 - - [18/Jul/2016:15:14:47 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:482016/07/18 15:14:48 [info] 8#8: *3251 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.224.205, server: localhost, request: "GET /api/heartbeat HTTP/1.0"
18/7/2016 17:14:4810.42.224.205 - - [18/Jul/2016:15:14:48 +0000] "GET /api/heartbeat HTTP/1.0" 200 38 "-" "-"
18/7/2016 17:14:482016/07/18 15:14:48 [info] 8#8: *3252 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.89.204, server: localhost, request: "GET /api/heartbeat HTTP/1.0"
```

**Expected:**

200 OK

**Comment by deniseschannon** [comment]
Can you share your docker-compose.yml and rancher-compose.yml of your service and your load balancer so I can try to recreate correctly?

I was unable to reproduce what I thought you were doing in v1.2.0-pre2

```
nginx:
  tty: true
  image: nginx:1.10.0
lbnginxtest:
  ports:
  - 80:80
  labels:
    io.rancher.scheduler.global: 'true'
  tty: true
  image: rancher/load-balancer-service
  links:
  - nginx:nginx
  stdin_open: true
```

rancher-compose.yml

```
nginx:
  scale: 8
lbnginxtest:
  load_balancer_config:
    haproxy_config: {}
  health_check:
    port: 42
    interval: 2000
    unhealthy_threshold: 3
    healthy_threshold: 2
    response_timeout: 2000
```

**Comment by suizman** [comment]
I'm using rancher server v1.1.0 not v1.2.0-pre2. Anyways my compose look like this

**docker-compose.yml**

```
nginx:
  image: my-nginx-with-custom-statics:1.10.0
lb:
  expose:
  - 80:80
  labels:
    io.rancher.scheduler.global: 'true'
  image: rancher/load-balancer-service
  links:
  - nginx:nginx
```

**rancher-compose.yml**
```

nginx:
  scale: 2
lb:
  scale: 2
  load_balancer_config:
    haproxy_config: {}
  health_check:
    port: 42
    interval: 2000
    unhealthy_threshold: 3
    healthy_threshold: 2
    response_timeout: 2000
```

On top of that I have external proxies pointing to the nginx service load balancers with public DNS..

I'm unable to reproduce the error in local it seams to be related when rancher server is under heavy load. I've detected that the network-agents are degradated and it coauses some serious conectivity issues with the services conected with links,rancher dns even if i try to ping the containers by IP.

Maybe the problems are more related with connectivity than haproxy and nginx itself ....

**Comment by deniseschannon** [comment]
@suizman Can you confirm if it has `expose` or `ports` for the load balancer?

**Comment by suizman** [comment]
Yes @deniseschannon i have **expose: 80:80** only for the rancher load balancer. No ports are declared.
BTW sry for the late respones ):

**Comment by roffe** [comment]
Rancher 1.1.4
CoreOS 1122.3
Docker 1.10.1 Swarm mode

Seeing the same issues on our relayd setup in openbsd

table <rancher_nodes> { 10.50.80.17, 10.50.80.18, 10.50.80.19 }

interval 2
log updates
prefork 10
timeout 200

redirect "rancher_http" {
        listen on ####### port 80
		forward to <rancher_nodes> check http "/" host health.rancher.check code 200
}


Then i have setup a loadbalancer in rancher going to 3 hello-world containers to just give 200 and response.

I get around 50% avail in the checks and in the logs i see:

relayctl show summary
Id      Type            Name                            Avlblty Status
1       redirect        rancher_http                            active
1       table           rancher_nodes:80                        active (2 hosts)
1       host            10.50.80.17                     40.78%  down
2       host            10.50.80.18                     48.54%  up
3       host            10.50.80.19                     54.37%  up

10/31/2016 5:50:34 PM2016/10/31 16:50:34 [info] 14#0: *8 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.19.174, server: , request: "HEAD / HTTP/1.1", host: "health.rancher.check"
10/31/2016 5:50:34 PM10.42.19.174 - - [31/Oct/2016:16:50:34 +0000] "HEAD / HTTP/1.1" 200 0 "-" "-"
10/31/2016 5:50:54 PM2016/10/31 16:50:54 [info] 14#0: *9 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.218.44, server: , request: "HEAD / HTTP/1.1", host: "health.rancher.check"
10/31/2016 5:50:54 PM10.42.218.44 - - [31/Oct/2016:16:50:54 +0000] "HEAD / HTTP/1.1" 200 0 "-" "-"
10/31/2016 5:51:24 PM2016/10/31 16:51:24 [info] 14#0: *10 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.218.44, server: , request: "HEAD / HTTP/1.1", host: "health.rancher.check"
10/31/2016 5:51:24 PM10.42.218.44 - - [31/Oct/2016:16:51:24 +0000] "HEAD / HTTP/1.1" 200 0 "-" "-"
10/31/2016 5:51:24 PM2016/10/31 16:51:24 [info] 14#0: *11 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.19.174, server: , request: "HEAD / HTTP/1.1", host: "health.rancher.check"
10/31/2016 5:51:24 PM10.42.19.174 - - [31/Oct/2016:16:51:24 +0000] "HEAD / HTTP/1.1" 200 0 "-" "-"
10/31/2016 5:51:44 PM2016/10/31 16:51:44 [info] 14#0: *12 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.120.224, server: , request: "HEAD / HTTP/1.1", host: "health.rancher.check"
10/31/2016 5:51:44 PM10.42.120.224 - - [31/Oct/2016:16:51:44 +0000] "HEAD / HTTP/1.1" 200 0 "-" "-"
10/31/2016 5:52:14 PM2016/10/31 16:52:14 [info] 14#0: *13 recv() failed (104: Connection reset by peer) while sending response to client, client: 10.42.218.44, server: , request: "HEAD / HTTP/1.1", host: "health.rancher.check"
10/31/2016 5:52:14 PM10.42.218.44 - - [31/Oct/2016:16:52:14 +0000] "HEAD / HTTP/1.1" 200 0 "-" "-"
10/31/2016 5:52:24 PM10.42.19.174 - - [31/Oct/2016:16:52:24 +0000] "HEAD / HTTP/1.1" 200 0 "-" "-"

<img width="850" alt="screen shot 2016-10-31 at 17 56 43" src="https://cloud.githubusercontent.com/assets/5434736/19863161/744c6e42-9f93-11e6-9e24-51597e9bca36.png">

**Comment by aemneina** [comment]
@suizman sounds like you've root caused this to a loaded rancher server. The 1.2.0 rancher server and agents are undergoing performance enhancements so this should be addressed in 1.2.0. Closing this out

**Comment by qvdk** [comment]
Hi, 
I have the same issue with rancher version 1.1.4 and 1.5.
The health check is working but generates a lot of 'connection reset by peer' into the application log.
